allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -289,14 +233,25 @@ def _read(self, file_path: str) -> Iterable[Instance]:
         raise NotImplementedError
 
     def _instances_from_cache_file(self, cache_filename: str) -> Iterable[Instance]:
+        # Try to acquire a lock just to make sure another process isn't in the middle
+        # of writing to the cache.
+        cache_file_lock = FileLock(cache_filename + ".lock", timeout=self.CACHE_FILE_LOCK_TIMEOUT)
```

Comment:
Putting the file locking logic inside this method makes it harder to override if you want better cache behavior.  Is it feasible to move the locking outside of this method?

Reply:
Yea, that's feasible. I was just thinking that someone at some point might want to use a non-local cache like Redis, at which point the lockfile would be meaningless. But that would probably require them to override other functionality as well.